File: dotnet/tests/Pactman.NonprofitCheckPlus.Tests/Support/RecordedRequest.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Pactman.NonprofitCheckPlus.Tests.Support;

/// <summary>One request that reached <see cref="FakeHandler"/>.</summary>
public sealed class RecordedRequest
{
    public RecordedRequest(string method, Uri uri, IReadOnlyDictionary<string, string> headers, string? body)
    {
        Method = method;
        Uri = uri;
        Headers = headers;
        Body = body;
    }

    public string Method { get; }

    public Uri Uri { get; }

    /// <summary>Every header sent, request and content headers together, keyed case-insensitively.</summary>
    public IReadOnlyDictionary<string, string> Headers { get; }

    public string? Body { get; }

    public string Path => Uri.AbsolutePath;

    public string? Header(string name) => Headers.TryGetValue(name, out var value) ? value : null;

    /// <summary>The request body parsed as JSON, or <see langword="null"/> when there was none.</summary>
    public JsonNode? JsonBody() => Body is null ? null : JsonNode.Parse(Body);

    /// <summary>The request body as a list of strings — the shape the bulk endpoint takes.</summary>
    public IReadOnlyList<string> StringListBody() =>
        Body is null ? Array.Empty<string>() : JsonSerializer.Deserialize<List<string>>(Body)!;
}

File: java/sdk/src/test/java/org/pactman/nonprofitcheckplus/support/RecordedRequest.java
package org.pactman.nonprofitcheckplus.support;

import java.net.URI;
import java.net.http.HttpHeaders;

/** One request {@link StubExchange} was asked to send. */
public final class RecordedRequest {

    private final URI uri;
    private final String method;
    private final HttpHeaders headers;
    private final String body;

    RecordedRequest(URI uri, String method, HttpHeaders headers, String body) {
        this.uri = uri;
        this.method = method;
        this.headers = headers;
        this.body = body;
    }

    /**
     * The URL the request was sent to.
     *
     * @return the URI.
     */
    public URI uri() {
        return uri;
    }

    /**
     * The HTTP method.
     *
     * @return the method.
     */
    public String method() {
        return method;
    }

    /**
     * One header's value.
     *
     * @param name the header name, case-insensitive.
     * @return the first value, or {@code null} when the header was not sent.
     */
    public String header(String name) {
        return headers.firstValue(name).orElse(null);
    }

    /**
     * The request body.
     *
     * @return the body, or {@code null} when the request had none.
     */
    public String body() {
        return body;
    }
}
